compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -566,7 +566,19 @@ impl<'ll, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'_, 'll, 'tcx> {
     }
 
     fn assume(&mut self, val: Self::Value) {
-        self.call_intrinsic("llvm.assume", &[val]);
+        match self.const_to_opt_uint(val) {
+            Some(0) => {
+                self.store(
```

Comment:
It doesn't look like this is documented. I'll find it or add it.

Reply:
> request: can you please include a comment here linking to something in LLVM about why this is worth doing? Otherwise someone like me would probably come along later and go "well, `llvm.assume` isn't a terminator so we can just emit that all the time instead" and try to "clean it up".
> 
> (Ideally there'd be something in https://llvm.org/docs/ with a bunch of these preferred forms that we could just link to, but no need to block this PR on that.)

See https://github.com/llvm/llvm-project/pull/98910.

Reply:
> For this PR if it ends up just being to a link to a comment from nikic or something that's fine -- wouldn't be the first time we have one of those in cg_llvm :P
> 
> But if you were able to get an official place in LLVM with these kinds of things, that'd be wonderful. I'd love to have a place to reference for all the "don't emit the unsigned checked math intrinsics directly" and such advice!

I'm guessing this is what you want? https://releases.llvm.org/18.1.0/docs/Frontend/PerformanceTips.html#id10

> Avoid using arithmetic intrinsics unless you are required by your source language specification to emit a particular code sequence. The optimizer is quite good at reasoning about general control flow and arithmetic, it is not anywhere near as strong at reasoning about the various intrinsics. If profitable for code generation purposes, the optimizer will likely form the intrinsics itself late in the optimization pipeline. It is very rarely profitable to emit these directly in the language frontend. This item explicitly includes the use of the [overflow intrinsics](https://releases.llvm.org/10.0.0/docs/LangRef.html#int-overflow).